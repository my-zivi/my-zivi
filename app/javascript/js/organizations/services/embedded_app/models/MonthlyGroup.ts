import { Moment } from 'moment';
import { DATE_FORMATS } from 'js/constants';
import { Service } from 'js/organizations/services/embedded_app/types';

export default class MonthlyGroup {
  public readonly services: Service[];
  public readonly month: Moment;

  constructor(month: Moment, services: Service[]) {
    this.services = services;
    this.month = month;
  }

  get monthName() {
    return this.month.format(DATE_FORMATS.fullMonth);
  }
}
