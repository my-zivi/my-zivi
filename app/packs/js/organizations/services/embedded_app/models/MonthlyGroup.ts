import { Moment } from 'moment';
import { DATE_FORMATS } from 'js/constants';

export default class MonthlyGroup {
  public readonly month: Moment;

  constructor(month: Moment) {
    this.month = month;
  }

  get monthName(): string {
    return this.monthBeginning.format(DATE_FORMATS.fullMonth);
  }

  get monthBeginning(): Moment {
    return Object.freeze(this.month.clone().startOf('month'));
  }

  get monthEnd(): Moment {
    return Object.freeze(this.month.clone().endOf('month'));
  }

  mapDays<T>(callback: (day: Moment) => T): T[] {
    const daysSpan = this.monthEnd.diff(this.monthBeginning, 'days');

    const output: Array<T> = [];
    for (let delta = 0; delta <= daysSpan; delta += 1) {
      output.push(callback(this.monthBeginning.add(delta, 'days')));
    }

    return output;
  }
}
