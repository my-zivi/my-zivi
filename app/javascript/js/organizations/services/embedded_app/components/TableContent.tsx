import React from 'preact/compat';
import { FunctionalComponent } from 'preact';
import MonthlyGroup from 'js/organizations/services/embedded_app/models/MonthlyGroup';

interface Props {
  monthlyGroup: MonthlyGroup;
}

const TableContent: FunctionalComponent<Props> = ({ monthlyGroup }) => (
  <>
    {monthlyGroup.services.map((service) => (
      <div className="d-table-row">
        <div className="d-table-cell">
          {service.civilServant.fullName}
        </div>
        {monthlyGroup.mapDays((day) => (
          <div className="d-table-cell">
            {day.isBetween(service.beginning, service.ending) ? 'x' : ''}
          </div>
        ))}
      </div>
    ))}
  </>
);

export default TableContent;
