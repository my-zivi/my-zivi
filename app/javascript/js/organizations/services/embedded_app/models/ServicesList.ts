import moment, { Moment } from 'moment';
import { range } from 'lodash';
import MonthlyGroup from 'js/organizations/services/embedded_app/models/MonthlyGroup';
import { Service } from 'js/organizations/services/embedded_app/types';

export default class ServicesList {
  public readonly services: Service[];
  public readonly earliestBeginning: Moment;
  public readonly latestEnding: Moment;
  private _monthlyGroups: MonthlyGroup[];

  constructor(services) {
    this.services = services;
    this.earliestBeginning = moment.min(this.services.map(service => moment(service['beginning'])));
    this.latestEnding = moment.max(this.services.map(service => moment(service['ending'])));
  }

  get planBeginning() {
    return moment(this.earliestBeginning).startOf('month');
  }

  get planEnding() {
    return moment(this.latestEnding).endOf('month');
  }

  get monthlyGroups() {
    if (this._monthlyGroups) {
      return this._monthlyGroups;
    }

    this._monthlyGroups = this.groupByMonth();
    return this._monthlyGroups;
  }

  daysSpan() {
    return this.planEnding.diff(this.planBeginning, 'days');
  }

  mapDays(callback: (day: Moment) => void) {
    let output = [];
    for (let delta = 0; delta <= this.daysSpan(); ++delta) {
      output.push(callback(this.planBeginning.add(delta, 'days')));
    }

    return output;
  }

  private groupByMonth() {
    const monthsCount = Math.ceil(this.planEnding.diff(this.planBeginning, 'months', true));
    const months = range(monthsCount).map(offset => {
      return this.planBeginning.add(offset, 'months').startOf('month');
    });

    return months.map(month => {
      return new MonthlyGroup(month,
        this.services.filter(service => {
          return moment(service.beginning).isSameOrBefore(month, 'month') || moment(service.ending).isSameOrAfter(month, 'month');
        }),
      );
    });
  }
}
