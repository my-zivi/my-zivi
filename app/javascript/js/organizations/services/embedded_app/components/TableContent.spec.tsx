import { shallow } from 'enzyme';
import moment from 'moment';
import React from 'preact/compat';
import ServicesList from '../models/ServicesList';
import Factories from '../../../../tests/factories/Factories';
import ServiceFactory from '../../../../tests/factories/ServiceFactory';
import MonthlyGroup from '../models/MonthlyGroup';
import TableContent from './TableContent';

describe('TableContent', () => {
  const servicesList = new ServicesList([
    ...Factories.buildList(ServiceFactory, 2),
    ServiceFactory.build({
      definitive: false,
      beginning: '2020-02-03',
      ending: '2020-03-06',
    }),
  ]);
  const monthlyGroup = new MonthlyGroup(moment(servicesList.services[0].beginning));
  const wrapper = shallow(<TableContent servicesList={servicesList} currentMonthlyGroup={monthlyGroup} />);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
