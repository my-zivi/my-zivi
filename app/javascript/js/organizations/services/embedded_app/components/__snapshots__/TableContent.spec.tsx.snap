// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableContent renders correctly 1`] = `
Array [
  <div
    className="d-table-row"
  >
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
  </div>,
  <div
    className="d-table-row"
  >
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell inactive"
    />
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
    <div
      className="d-table-cell day-cell active"
    >
      <i
        className="fas fa-check"
      />
    </div>
  </div>,
]
`;
